integration/backfill_test.go

```diff
@@ -0,0 +1,280 @@
+		verifyBlock(t, client, block2, block2Path)
+	}
+
+	{
+		// Let's try to upload broken block (no meta.json)
+		badBlock, err := testhelper.CreateBlock(context.Background(), tmpDir, []labels.Labels{labels.FromStrings("test", "bad")}, 100, blockEnd.Add(-2*time.Hour).UnixMilli(), blockEnd.UnixMilli(), nil, 0, metadata.NoneFunc)
+		require.NoError(t, err)
+		badBlockPath := filepath.Join(tmpDir, badBlock.String())
+
+		require.NoError(t, os.Remove(filepath.Join(badBlockPath, block.MetaFilename)))
+
+		output, err := runMimirtoolBackfill(t, "--address", compactorURL, "--id", "anonymous", badBlockPath)
+		require.Regexp(t, fmt.Sprintf("msg=\"failed uploading block\"[^\n]+path=%s", badBlockPath), output)
+		require.Error(t, err)
+	}
+
+	{
+		// Let's try block with external labels. Mimir currently rejects those.
+		extLabels := labels.FromStrings("ext", "labels")
+
+		badBlock, err := testhelper.CreateBlock(context.Background(), tmpDir, []labels.Labels{labels.FromStrings("test", "bad")}, 100, blockEnd.Add(-2*time.Hour).UnixMilli(), blockEnd.UnixMilli(), extLabels, 0, metadata.NoneFunc)
+		require.NoError(t, err)
+		badBlockPath := filepath.Join(tmpDir, badBlock.String())
+
+		output, err := runMimirtoolBackfill(t, "--address", compactorURL, "--id", "anonymous", badBlockPath)
+		require.Contains(t, output, "unsupported external label: ext")
+		require.Error(t, err)
+	}
+
+}
+
+func verifyBlock(t *testing.T, client objstore.Bucket, ulid ulid.ULID, localPath string) {
+	for _, fn := range []string{"index", "chunks/000001"} {
+		a, err := client.Attributes(context.Background(), path.Join("blocks/anonymous", ulid.String(), fn))
+		require.NoError(t, err)
+
+		fi, err := os.Stat(filepath.Join(localPath, filepath.FromSlash(fn)))
+		require.NoError(t, err)
+
+		require.Equal(t, a.Size, fi.Size())
+	}
+
+	localMeta, err := metadata.ReadFromDir(localPath)
+	require.NoError(t, err)
+
+	remoteMeta, err := block.DownloadMeta(context.Background(), log.NewNopLogger(), bucket.NewPrefixedBucketClient(client, "blocks/anonymous"), ulid)
+	require.NoError(t, err)
+
+	require.Equal(t, localMeta.ULID, remoteMeta.ULID)
+	require.Equal(t, localMeta.Stats, remoteMeta.Stats)
+}
+
+func runMimirtoolBackfill(t *testing.T, args ...string) (string, error) {
+	oldStdout, oldStderr := os.Stdout, os.Stderr
+	defer func() {
+		os.Stdout = oldStdout
+		os.Stderr = oldStderr
+
+		logrus.SetOutput(os.Stderr)
+	}()
+
+	co := captureOutput(t)
+	// mimirtool uses logrus, which has global logger. We reinitialize it to use our captured Stderr.
+	logrus.SetOutput(os.Stderr)
+
+	app := kingpin.New("mimirtool", "help")
+	envVars := commands.NewEnvVarsWithPrefix("MIMIR")
+
+	// Only register mimirtool backfill command.
+	backfill := commands.BackfillCommand{}
+	backfill.Register(app, envVars)
+
+	// Run mimirtool (Parse calls any action).
+	_, err := app.Parse(append([]string{"backfill"}, args...))
+
+	stdout, stderr := co.Done()
+	return stdout + stderr, err
+}
+
+type capturedOutput struct {
```

Comment:
Looks a duplicate of `cmd/mimir/main_test.go`. I would move it to a common util package.